samples/logging/prometheus-grafana-opentelemetry/sample.md

```diff
@@ -0,0 +1,134 @@
+
+snippet: enable-opentelemetry-metrics
+
+There are three metrics reported as a Counter:
+
+ * Number of fetched messages 
+ * Number of failed messages
+ * Number of successfully processed messages
+
+Each reported metric is tagged with the following additional information:
+
+ * `messaging_discriminator` a uniquely addressable address for the endpoint (discriminator when scaling out)
+ * `messaging_endpoint` endpoint name
+ * `messaging_queue` the queue name of the endpoint
+ * `messaging_type` the .NET type information for the message being processed
+ * `messaging_failure_type` the exception type name (if applicable)
+
+## Exporting metrics
+
+The metrics are gathered using OpenTelemetry standard on the endpoint and need to be reported and collected by an external service. A Prometheus exporter can expose this data to the Prometheus service, hosted as a docker service. This exporter is available via a NuGet package `OpenTelemetry.Exporter.Prometheus`. The sample exposes an HTTP endpoint that enables Prometheus to scrape data reported by the OpenTelemetry metrics. In the sample the service that exposes the data to scrape is hosted on `http://localhost:9185/metrics`. To enable the Prometheus exporter, the following should be executed:
+
+snippet: enable-prometheus-exporter
+
+Note: that the HTTP endpoint is also exposed through a local IP address so the Prometheus on docker can reach.
+
+The raw metrics retrieved through the scraping endpoint would look like this:
+
+```text
+# HELP messaging_successes Total number of messages processed successfully by the endpoint.
+# TYPE messaging_successes counter
+messaging_successes{messaging_discriminator="main",messaging_endpoint="Samples.OpenTelemetry.Metrics",messaging_queue="Samples.OpenTelemetry.Metrics",messaging_type="SomeCommand, Otel.MetricsDemo, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null"} 526 1656039256653
+
+# HELP messaging_fetches Total number of messages fetched from the queue by the endpoint.
+# TYPE messaging_fetches counter
+messaging_fetches{messaging_discriminator="main",messaging_endpoint="Samples.OpenTelemetry.Metrics",messaging_queue="Samples.OpenTelemetry.Metrics",messaging_type="SomeCommand, Otel.MetricsDemo, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null"} 643 1656039256653
+
+# HELP messaging_failures Total number of messages processed unsuccessfully by the endpoint.
+# TYPE messaging_failures counter
+messaging_failures{messaging_discriminator="main",messaging_endpoint="Samples.OpenTelemetry.Metrics",messaging_failure_type="System.Exception",messaging_queue="Samples.OpenTelemetry.Metrics",messaging_type="SomeCommand, Otel.MetricsDemo, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null"} 117 1656039256653
+```
+
+Custom observers need to be registered for the metric probes provided via `NServiceBus.Metrics`. This is all setup in the `PrometheusFeature`
```

Comment:
Is there sample doing this? It sounds interesting but maybe we should link to the other sample to keep this one focused on getting OpenTelemetry data exported.

Reply:
Cleaned up the sample and removed irrelevant parts.